### ISSUE TITLE
Question about LoRA implementation when loading Llama2 checkpoint

### ISSUE THREAD

**Issue by ebsmothers** [Lightning-AI/lit-gpt#891]
Hi there, 

I have a question about the way that you are applying LoRAQKVLinear to the Llama2 model. If I understand the [checkpoint remapping](https://github.com/Lightning-AI/lit-gpt/blob/0f021f3ad8cd8d6fe30b0ef721a7a0e2dac15898/scripts/convert_hf_checkpoint.py#L174-L187) correctly, the weights from Q, K, and V are interleaved along the embedding dimension.

Concretely suppose that I have 2 heads each having dimension 8 and I am using normal MHA. Then after loading the HF weights using that script my full QKV projection matrix should look like `[Q[:8], K[:8], V[:8], Q[8:], K[8:], V[8:]]`. Now, suppose I am only applying LoRA to the Q and V matrices. Then the embedding dim of my unpadded LoRA outputs [here](https://github.com/Lightning-AI/lit-gpt/blob/dbbbe3a6c09edfa2fdd48a0bc997d51cc105c98a/lit_gpt/lora.py#L407-L412) would be 16, which need to be padded to 24. But the [padding method](https://github.com/Lightning-AI/lit-gpt/blob/dbbbe3a6c09edfa2fdd48a0bc997d51cc105c98a/lit_gpt/lora.py#L274-L295) just does a simple concatenation of Q, K, and V (i.e. no interleaving).

So if we denote `Q1 = Q[:8]`, `Q2 = Q[8:]`, `BAQ1` and `BAQ2` the analogous slices of Q's LoRA decomposition, and similar notation for K and V, we wind up adding the two matrices
```
[Q1, K1, V1, Q2, K2, V2] + [BAQ1, BAQ2, BAK1, BAK2, BAV1, BAV2]
```

You can see that only in the first and last blocks will we actually have the desired LoRA decomposition, but all others will include cross terms. This seems surprising to me. Is there something obvious I'm missing in the checkpoint loading or finetuning script? 

Thanks in advance for your assistance and thanks for your great repo!

**Comment by Andrei-Aksionov** [comment]
Hello @ebsmothers, thanks for a wonderful question (and for analysis too)!

I see where the confusion comes from.
In the [remapping](https://github.com/Lightning-AI/lit-gpt/blob/0f021f3ad8cd8d6fe30b0ef721a7a0e2dac15898/scripts/convert_hf_checkpoint.py#L150-L188) code we first iterate over layers and "accumulate" `q`, `k` and `v` matrices, so we can merge weights and assign it to a combined QKV matrix of a particular layer.
So `i` value in the loop means the layer number, not the position of "interleaving", like you denoted
> [Q[:8], K[:8], V[:8], Q[8:], K[8:], V[8:]]

For clarification, let's unwrap how weights are stored in the combined QKV matrix.
QKV matrix consists of three matrices (query, key and value), each matrix dimension consists of `n_head` * `n_emd`.
So the structure of 2 heads of 8 dim would be like this:
`[Q[:8], Q[:8], K[:8], K[8:], V[8:], V[8:]]`.
If you prefer to read the code with annotations, I have added them [here](https://github.com/Lightning-AI/lit-llama/blob/da71adea0970d6d950fb966d365cfb428aef8298/lit_llama/adapter.py#L105-L126) (regular MHA). Haven't yet ported them into this repo, oops 😊.

**Comment by Andrei-Aksionov** [comment]
Hello @ebsmothers, thanks for a wonderful question (and for analysis too)!

Looks like you are right. Nice catch 👍 
Apparently Q, K and V weights became interleaved after the work to support grouped queries, but LoRA code hasn't been updated accordingly.

**Comment by Andrei-Aksionov** [comment]
At the first glance, looks like it can be fixed by just changing `self.lora_ind`, so it reflects the correct structure.
@ebsmothers would you like to work on it and create a PR?

**Comment by carmocca** [comment]
Pinging again. Is one of you interested in working on a fix? I would expect the fix to have similar interleaving logic as in the ckpt conversion: https://github.com/Lightning-AI/lit-gpt/blob/0f021f3ad8cd8d6fe30b0ef721a7a0e2dac15898/scripts/convert_hf_checkpoint.py#L181-L186

Interestingly, we've fine-tuned llama 2 models with LoRA and the model does learn. The impact of this bug is not clear

**Comment by ebsmothers** [comment]
Hi @Andrei-Aksionov and @carmocca apologies for the delayed response. I just put up #900 as a proof of concept here. Tbh I do not have a ton of bandwidth at the moment and have not tested the changes yet, so if the fix is time-sensitive at all feel free to commandeer to move things along.

**Comment by Andrei-Aksionov** [comment]
> Interestingly, we've fine-tuned llama 2 models with LoRA and the model does learn. The impact of this bug is not clear

The LoRA fine-tuning did work because the code managed to learn weight updates (albeit it learned weight updates for the wrong parts of the QKV matrix 🙃).

From the example provided by @ebsmothers:
```text
[Q1, K1, V1, Q2, K2, V2] + [BAQ1, BAQ2, BAK1, BAK2, BAV1, BAV2]
```

If we disable LoRA for `keys`, but keep for `queries` and `values`, it will be:
```text
[Q1, K1, V1, Q2, K2, V2] + [BAQ1, BAQ2, 0, 0, BAV1, BAV2]
```

which means that `Q1`, `K1`, `K2` and `V2` are updated.

Since LoRA matrices are initialized identically for different parts of the QKV matrix, that allows such a confusion.

I am interested to see what the metrics are after this fix is applied.
